src/python/pants/backend/python/providers/pyenv/rules.py

```diff
@@ -0,0 +1,336 @@
+from pants.engine.fs import CreateDigest, FileContent
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.internals.synthetic_targets import SyntheticAddressMaps, SyntheticTargetsRequest
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessCacheScope, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.util.docutil import bin_name
+from pants.util.frozendict import FrozenDict
+from pants.util.meta import classproperty
+from pants.util.strutil import softwrap
+
+PYENV_NAMED_CACHE = ".pyenv"
+PYENV_APPEND_ONLY_CACHES = FrozenDict({"pyenv": PYENV_NAMED_CACHE})
+
+
+class PyenvSubsystem(TemplatedExternalTool):
+    options_scope = "pyenv"
+    name = "pyenv"
+    help = softwrap(
+        f"""
+        A subsystem for pyenv-provided-Python (https://github.com/pyenv/pyenv).
+
+        Enabling this subsystem will switch Pants from trying to find an appropriate Python on your
+        system to using pyenv to install the correct Python(s).
+
+        The Pythons provided by Pyenv will be used to run any "user" code (your Python code as well
+        as any Python-based tools you use, like black or pylint). The Pythons are also read-only to
+        ensure they remain hermetic across runs of different tools and code.
+
+        The Pythons themselves are stored in your `named_caches_dir`: https://www.pantsbuild.org/docs/reference-global#named_caches_dir
+        under `pyenv/versions/<version>`. Wiping the relevant version directory (with `sudo rm -rf`)
+        will force a re-install of Python. This may be necessary after changing something about the
+        underlying system which changes the compiled Python, such as installing an
+        optional-at-build-time dependency like `liblzma-dev` (which is used for the optional module
+        `lzma`).
+
+        By default, the subsystem does not pass any optimization flags to the Python compilation
+        process. Doing so would increase the time it takes to install a single Python by about an
+        order of magnitude (E.g. ~2.5 minutes to ~26 minutes).
+
+        If you wish to customize the pyenv installation of a Python, a synthetic target is exposed
+        at the root of your repo which is runnable. This target can be run with the relevant
+        environment variables set to enable optimizations. You will need to wipe the specific
+        version directory if Python was already installed. Example:
+
+            sudo rm -rf <named_caches_dir>/pyenv/versions/<specific_version>
+            # env vars from https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-for-maximum-performance
+            PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto' {bin_name()} run :pyenv-install -- 3.10
```

Comment:
This is cool. We effectively get "free" sub-commands to our goals with synthetic targets!

Reply:
your last commit made this comment stale.

Reply:
fixed